### PR TITLE
JDK-8289551: Conversions between bit representations of half precision values and floats

### PR DESCRIPTION
**Pull request by jddarcy**
Initial implementation.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8289551](https://bugs.openjdk.org/browse/JDK-8289551): Conversions between bit representations of half precision values and floats
 * [JDK-8290216](https://bugs.openjdk.org/browse/JDK-8290216): Conversions between bit representations of half precision values and floats (**CSR**)

### Reviewers
 * [Paul Sandoz](https://openjdk.org/census#psandoz) (@PaulSandoz - **Reviewer**) ⚠️ Review applies to [d54f9c94](https://git.openjdk.org/jdk/pull/9422/files/d54f9c948369d11db3b72e7fbf4db8981e66ddf9)
 * [John R Rose](https://openjdk.org/census#jrose) (@rose00 - **Reviewer**) ⚠️ Review applies to [d54f9c94](https://git.openjdk.org/jdk/pull/9422/files/d54f9c948369d11db3b72e7fbf4db8981e66ddf9)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9422/head:pull/9422` \
`$ git checkout pull/9422`

Update a local copy of the PR: \
`$ git checkout pull/9422` \
`$ git pull https://git.openjdk.org/jdk pull/9422/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9422`

View PR using the GUI difftool: \
`$ git pr show -t 9422`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9422.diff">https://git.openjdk.org/jdk/pull/9422.diff</a>

</details>

### CODE REVIEW

**Comment by jddarcy** [comment]
Some notes in the current implementation in the PR: the impetus for the change is to provide a minimal level of platform support for the binary16 floating-point format defined in IEEE 754. This is done by added two methods to Float, one for binary16 -> float conversion and the other for float -> binary16 conversion. In the absence of the ability to define primitive types, short is used as a carrier for the bits of a binary16 value.

The conversion code in its current form favors readability over speed; a more performant software implementation may be possible even without intrinsification.

A different 16-bit floating-point format, bfloat, is popular in some circles. Any similar support for bfloat will be left for future work.

Please also review the companion CSR: https://bugs.openjdk.org/browse/JDK-8290216

File: src/java.base/share/classes/java/lang/Float.java [hunk 1]
```diff
@@ -975,6 +976,206 @@ public static int floatToIntBits(float value) {
...
+                //
+                // The exponent range of normalized binary16 subnormal
+                // values is [-15, -24]. The exponent range of float
+                // subnormals is [-140, -149]. Multiply abs_f down by
+                // 2^(-125) -- since (-125 = -149 - (-24)) -- so that
+                // the trailing bits of a subnormal float represent
+                // the correct trailing bits of a binary16 subnormal.
+                exp = -15; // Subnormal encoding using -E_max.
+                float f_adjust = abs_f * 0x1.0p-125f;
+                signif_bits = (short)(Float.floatToRawIntBits(f_adjust) & 0x03ff);
```
Comment (rgiulietti) [inline src/java.base/share/classes/java/lang/Float.java:1127]: I think the `if` and the `exp++` can be avoided if the `& 0x03ff` is dropped altogether.
```
               signif_bits = (short)Float.floatToRawIntBits(f_adjust);
```
The reason is the same as for the normalized case below: a carry will eventually flow into the representation for the exponent.

File: test/jdk/java/lang/Float/SixteenBitFormats.java [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+                    if (isNaN(bin16_2)) {
+                        return -1;
+                    }
+                    return Integer.compare((int)(bin16_1 & 0xFFFF),
+                                           (int)(bin16_2 & 0xFFFF));
+                }
+            }
+        }
+    }
+}
```
Comment (rgiulietti) [inline test/jdk/java/lang/Float/SixteenBitFormats.java:263]: The last invocation to `Integer.compare()` is not correct.
For example, if `bin16_1 = -1 `and `bin16_2 = 1`, the invocation returns 1, which is incorrect.

File: src/java.base/share/classes/java/lang/Float.java [hunk 3]
```diff
@@ -975,6 +976,206 @@ public static int floatToIntBits(float value) {
...
+                    // to be adjusted accordingly.
+                    exp++;
+                }
+            } else {
+                // All remaining values of f are in the normalized
+                // range of binary16 (which is also in the normalized
+                // range of float).
+
+                // Significand bits as if using rounding to zero (truncation).
+                signif_bits = (short)((doppel & 0x0007f_e000) >>
```
[the remark this answers is not in the document]
Comment (jddarcy) [reply src/java.base/share/classes/java/lang/Float.java:1141]: Oops; yes, I intended for the int hex constants to have at most 8 hex digits.

File: test/jdk/java/lang/Float/SixteenBitFormats.java [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+        public static final short MAX_SUBNORMAL = (short)0x03ff;
+
+        public static boolean isNaN(short binary16) {
+            return ((binary16 & 0x7c00) == 0x7c00) // Max exponent and...
+                && ((binary16 & 0x03ff) != 0 );    // significand nonzero.
+        }
+
+        public static short negate(short binary16) {
+            return (short)(((binary16 & 0x8000) ^ 0x8000) | // Isolate and flip sign bit
+                           (binary16 & 0x7fff));
```
[the remark this answers is not in the document]
Comment (jddarcy) [reply test/jdk/java/lang/Float/SixteenBitFormats.java:244]: Good suggestion.

File: test/jdk/java/lang/Float/SixteenBitFormats.java [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+        public static final short ONE = 0x3c00;
+        public static final short MIN_NORMAL = 0x0400;
+        public static final short MIN_VALUE = 0x0001;
+        public static final short NEGATIVE_ZERO = (short)0x8000;
+        public static final short POSITIVE_ZERO = 0x0000;
+        public static final short MAX_SUBNORMAL = (short)0x03ff;
+
+        public static boolean isNaN(short binary16) {
+            return ((binary16 & 0x7c00) == 0x7c00) // Max exponent and...
+                && ((binary16 & 0x03ff) != 0 );    // significand nonzero.
```
[the remark this answers is not in the document]
Comment (jddarcy) [reply test/jdk/java/lang/Float/SixteenBitFormats.java:239]: In this case, I think the more verbose coding pattern is clearer.

File: src/java.base/share/classes/java/lang/Float.java [hunk 6]
```diff
@@ -975,6 +976,206 @@ public static int floatToIntBits(float value) {
...
+     * The operation of this method is analogous to a primitive
+     * narrowing conversion (JLS {@jls 5.1.3}).
+     *
+     * @param f the {@code float} value to convert to binary16
+     * @since 20
+     */
+    // @IntrinsicCandidate
+    public static short floatToBinary16AsShortBits(float f) {
+        if (Float.isNaN(f)) {
+            // Arbitrary binary16 NaN value; could try to preserve the
```
[the remark this answers is not in the document]
Comment (jddarcy) [reply src/java.base/share/classes/java/lang/Float.java:1090]: I double-checked the 754-2019 standard; unlike in some earlier versions of the standard, the 2019 edition does specify how a quiet vs signaling NaN is supposed to be encoded. Quoting from section 6.2.1:

"A quiet NaN bit string should be encoded with the first bit (d1) of the trailing significand
field T being 1. A signaling NaN bit string should be encoded with the first bit of the trailing significand
field being 0."

So the NaN bit pattern used is a quiet NaN.

In any case, I'll update this code to more faithfully pass along the bits of a Float NaN.

File: src/java.base/share/classes/java/lang/Float.java [hunk 7]
```diff
@@ -975,6 +976,206 @@ public static int floatToIntBits(float value) {
...
+                    ( lsb != 0  &&  round != 0 ) ) { // sticky not needed
+                    // Due to the representational properties, an
+                    // increment will also handle a ripple carry that
+                    // updates the exponent too.
+                    signif_bits++;
+                }
+            }
+
+            short result = 0;
+            result = (short)(((exp + 15) << 10) | signif_bits);
```
[the remark this answers is not in the document]
Comment (jddarcy) [reply src/java.base/share/classes/java/lang/Float.java:1174]: Good catch; I'll correct this bug in the next push.

I'll be adding a test to verify correct rounding of all the values in a binade - the set of floating-point values with the same exponent. That test reveals this issue.

File: src/java.base/share/classes/java/lang/Float.java [hunk 8]
```diff
@@ -975,6 +976,215 @@ public static int floatToIntBits(float value) {
...
+                // 1    1     1
+                // See "Computer Arithmetic Algorithms," Koren, Table 4.9
+
+                // Bits of binary16 significand in a float: 0x0007f_e000;
+                // therefore, the other quantities of interest are:
+                int lsb   =  doppel & 0x0000_2000;
+                int round =  doppel & 0x0000_1000;
+                int sticky = doppel & 0x0000_0fff;
+
+                if (round != 0 && (lsb != 0 || sticky != 0 )) {
```
Comment (rgiulietti) [inline src/java.base/share/classes/java/lang/Float.java:1173]: You may want to reverse the order of `(lsb != 0 || sticky != 0)` to `(sticky != 0 || lsb != 0)` to improve the likelihood that the evaluation of `||` does not need to evaluate its right-hand side.
Statistically, `sticky != 0` has a much higher chance to evaluate to `true` than `lsb != 0`.

File: src/java.base/share/classes/java/lang/Float.java [hunk 9]
```diff
@@ -975,6 +976,215 @@ public static int floatToIntBits(float value) {
...
+                // 2^(-125) -- since (-125 = -149 - (-24)) -- so that
+                // the trailing bits of a subnormal float represent
+                // the correct trailing bits of a binary16 subnormal.
+                exp = -15; // Subnormal encoding using -E_max.
+                float f_adjust = abs_f * 0x1.0p-125f;
+
+                // In case the significand rounds up and has a carry
+                // propagate all the way up, take the bottom 11 bits
+                // rather than bottom 10 bits. Adding this value,
+                // rather than OR'ing htis value, will cause the right
```
Comment (rgiulietti) [inline src/java.base/share/classes/java/lang/Float.java:1141]: `htis` -> `this`

**Comment by jddarcy** [comment]
Thanks for the review comments thus far.

A few notes on the recently pushed update, corresponding to webrev 04.

The conversion methods have been renamed as suggested in comments received outside of the PR.

Thinking more about how a future possible value class for the binary16 format might be structured (java.math.FloatBinary16? java.math.Binary16?), even if such a class cannot extend java.lang.Number, it could still implement the conversion methods found on Number. In particular, a value class for binary16 could have a "floatValue" method as well as factory to convert from float to binary16. Therefore, there need not be any overloading concerns with the methods defined on Float for this issue.

(As an aside, it might be helpful to define a superinterface with the methods of java.lang.Number to better capture the ConvertibleToPrimitive API.)

The regression tests were augmented in several ways. One is by the introduction of an alternative implementation of the float -> binary16 conversion that uses the hardware support for float rounding to compute the proper low-order bits. This approach benchmarked as significantly slower than the existing bit-inspection approach so it is just used to double-check in testing. Before updating the PR, all 2^32 float values were converted to binary16 using both approaches and the results matched. Testing all 2^32 values all the time seems excessive, so in the PR just one exponent value is checked. For each exponent in normal range, the rounding is analagous at corresponding positions of the number line. Therefore, testing one exponent should be sufficient for testing all the different combination of bit that influence rounding.

File: test/jdk/java/lang/Float/Binary16Conversion.java [hunk 10]
```diff
@@ -0,0 +1,429 @@
...
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8289551
+ * @summary Verify conversion between float and the binary16 format
+ * @library ../Math
+ * @build FloatConsts
```
Comment (PaulSandoz) [inline test/jdk/java/lang/Float/Binary16Conversion.java:29]: Is this line necessary?

File: test/jdk/java/lang/Float/Binary16Conversion.java [hunk 11]
```diff
@@ -0,0 +1,429 @@
...
+ * @run main Binary16Conversion
+ */
+
+import static java.lang.Float.*;
```
Comment (PaulSandoz) [inline test/jdk/java/lang/Float/Binary16Conversion.java:33]: Unused.

File: test/jdk/java/lang/Float/Binary16Conversion.java [hunk 12]
```diff
@@ -0,0 +1,429 @@
...
+        }
+
+        // float -> short
+        for (var testCase : testCases) {
+            errors += compareAndReportError(testCase[1],
+                                            (short)testCase[0]);
+        }
+
+        return errors;
+    };
```
Comment (PaulSandoz) [inline test/jdk/java/lang/Float/Binary16Conversion.java:100]:
```suggestion
    }
```

File: test/jdk/java/lang/Float/Binary16Conversion.java [hunk 13]
```diff
@@ -0,0 +1,429 @@
...
+
+    public static class Binary16 {
+        public static final short POSITIVE_INFINITY = (short)0x7c00;
+        public static final short NEGATIVE_INFINITY = (short)0xfc00;
+        public static final short MAX_VALUE         = 0x7bff;
+        public static final short ONE               = 0x3c00;
+        public static final short MIN_NORMAL        = 0x0400;
+        public static final short MAX_SUBNORMAL     = 0x03ff;
+        public static final short MIN_VALUE         = 0x0001;
+        public static final short NEGATIVE_ZERO     = (short)0x8000;
```
Comment (PaulSandoz) [inline test/jdk/java/lang/Float/Binary16Conversion.java:409]: Unused.

File: test/jdk/java/lang/Float/Binary16ConversionNaN.java [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8289551
+ * @summary Verify NaN sign and significand bits are preserved across conversions
+ */
+
+import static java.lang.Float.*;
```
Comment (PaulSandoz) [inline test/jdk/java/lang/Float/Binary16ConversionNaN.java:30]: Unused.

**Review by PaulSandoz — APPROVED** [review]
I don't claim to be an expert in IEEE formats... the methods generally look good with thorough testing. Looking forward to the impact intrinsification will have!

**Review by rose00 — APPROVED** [review]
Looks good.

File: src/java.base/share/classes/java/lang/Float.java [hunk 15]
```diff
@@ -975,6 +976,215 @@ public static int floatToIntBits(float value) {
...
+            // Smallest magnitude nonzero representable binary16 value
+            // is equal to 0x1.0p-24; half-way and smaller rounds to zero.
+            if (abs_f <= 0x1.0p-25f) { // Covers float zeros and subnormals.
+                return sign_bit; // Positive or negative zero
+            }
+
+            // Dealing with finite values in exponent range of
+            // binary16 (when rounding is done, could still round up)
+            int exp = Math.getExponent(f);
+            assert -25 <= exp && exp <= 15;
```
[the remark this answers is not in the document]
Comment (rgiulietti) [reply src/java.base/share/classes/java/lang/Float.java:1122]: I didn't test this variant, will do tomorrow when also reviewing the tests themselves.

[the remark this answers is not in the document]
Comment (rgiulietti) [reply src/java.base/share/classes/java/lang/Float.java:1122]: The correct variant below passes the tests.

```
        // For binary16 subnormals, beside forcing exp to -15,
        // retain the difference expdelta = E_min - exp.
        // This is the excess shift value, in addition to 13, to be used
        // in the computations below.
        // Further the (hidden) msb with value 1 in f must be involved as well.
        int expdelta = 0;
        int msb = 0x0000_0000;
        if (exp < -14) {
            expdelta = -14 - exp;
            exp = -15;
            msb = 0x0080_0000;
        }
        int f_signif_bits = doppel & 0x007f_ffff | msb;

        // Significand bits as if using rounding to zero (truncation).
        short signif_bits = (short)(f_signif_bits >> (13 + expdelta));

        // For round to nearest even, determining whether or
        // not to round up (in magnitude) is a function of the
        // least significant bit (LSB), the next bit position
        // (the round position), and the sticky bit (whether
        // there are any nonzero bits in the exact result to
        // the right of the round digit). An increment occurs
        // in three cases:
        //
        // LSB  Round Sticky
        // 0    1     1
        // 1    1     0
        // 1    1     1
        // See "Computer Arithmetic Algorithms," Koren, Table 4.9

        int lsb    = f_signif_bits & (1 << 13 + expdelta);
        int round  = f_signif_bits & (1 << 12 + expdelta);
        int sticky = f_signif_bits & ((1 << 12 + expdelta) - 1);

        if (round != 0 && ((lsb | sticky) != 0 )) {
            signif_bits++;
        }

        // No bits set in significand beyond the *first* exponent
        // bit, not just the sigificand; quantity is added to the
        // exponent to implement a carry out from rounding the
        // significand.
        assert (0xf800 & signif_bits) == 0x0;

        return (short)(sign_bit | ( ((exp + 15) << 10) + signif_bits ) );
```

[the remark this answers is not in the document]
Comment (rgiulietti) [reply src/java.base/share/classes/java/lang/Float.java:1122]: Test classes look good.

File: src/java.base/share/classes/java/lang/Float.java [hunk 16]
```diff
@@ -975,6 +976,215 @@ public static int floatToIntBits(float value) {
     @IntrinsicCandidate
     public static native float intBitsToFloat(int bits);
 
+    /**
+     * {@return the {@code float} value closest to the numerical value
```
[the remark this answers is not in the document]
Comment (jddarcy) [reply src/java.base/share/classes/java/lang/Float.java:980]: The closing brace is on line 982, "... {@code short}}"

(Using the in-line form of the @return tag allows an explicit @return tag to be skipped by reusing the initial sentence. The generated javadoc is as intended.)

File: src/java.base/share/classes/java/lang/Float.java [hunk 17]
```diff
@@ -975,6 +976,215 @@ public static int floatToIntBits(float value) {
...
+                    | (doppel & 0x0000_1ff0) >> 4  //  9 bits
+                    | (doppel & 0x0000_000f));     //  4 bits
+        }
+
+        float abs_f = Math.abs(f);
+
+        // The overflow threshold is binary16 MAX_VALUE + 1/2 ulp
+        if (abs_f >= (65504.0f + 16.0f) ) {
+            return (short)(sign_bit | 0x7c00); // Positive or negative infinity
+        } else {
```
[the remark this answers is not in the document]
Comment (jddarcy) [reply src/java.base/share/classes/java/lang/Float.java:1112]: Good suggestion; the reduced indenting in this case also allows some of the comment blocks to take up fewer lines.

**Comment by rgiulietti** [comment]
LGTM

**Pull request closed without merging**